src/psbt.c

```diff
@@ -2201,3 +2203,733 @@ int wally_psbt_to_base64(
     }
     return ret;
 }
+
+static int get_txid(
+    struct wally_tx *tx,
+    unsigned char *txid,
+    size_t txid_len)
+{
+    unsigned char *bytes = NULL;
+    size_t calc, written;
+    int ret = WALLY_OK;
+
+    if (!tx || !txid || txid_len != SHA256_LEN) {
+        return WALLY_EINVAL;
+    }
+
+    if ((ret = wally_tx_get_length(tx, 0, &calc)) != WALLY_OK) {
+        return ret;
+    }
+    if ((bytes = wally_malloc(calc)) == NULL) {
+        return WALLY_ENOMEM;
+    }
+    if ((ret = wally_tx_to_bytes(tx, 0, bytes, calc, &written)) == WALLY_OK) {
+        ret = wally_sha256d(bytes, written, txid, SHA256_LEN);
+    }
+
+    wally_free(bytes);
+    return ret;
+}
+
+static int merge_input_into(
+    struct wally_psbt_input *dst,
+    struct wally_psbt_input *src)
+{
+    int ret = WALLY_OK;
+    size_t i, j;
+
+    if (!dst->non_witness_utxo && src->non_witness_utxo && (ret = clone_tx(src->non_witness_utxo, &dst->non_witness_utxo)) != WALLY_OK) {
+        return ret;
+    }
+    if (!dst->witness_utxo && src->witness_utxo && (ret = wally_tx_output_init_alloc(src->witness_utxo->satoshi, src->witness_utxo->script, src->witness_utxo->script_len, &dst->witness_utxo)) != WALLY_OK) {
+        return ret;
+    }
+
+    if (src->partial_sigs) {
+        if (!dst->partial_sigs) {
+            if ((ret = wally_partial_sigs_map_init_alloc(src->partial_sigs->items_allocation_len, &dst->partial_sigs)) != WALLY_OK) {
+                return ret;
+            }
+        }
+
+        for (i = 0; i < src->partial_sigs->num_items; ++i) {
+            bool found = false;
+            for (j = 0; j < dst->partial_sigs->num_items; ++j) {
+                if (memcmp((char *)dst->partial_sigs->items[j].pubkey, (char *)src->partial_sigs->items[i].pubkey, 65) == 0) {
+                    found = true;
+                    break;
+                }
+            }
+            if (found) {
+                continue;
+            }
+            if ((ret = add_partial_sig_item(dst->partial_sigs, &src->partial_sigs->items[i])) != WALLY_OK) {
+                return ret;
+            }
+        }
+    }
+
+    if (src->keypaths) {
+        if (!dst->keypaths) {
+            if ((ret = wally_keypath_map_init_alloc(src->keypaths->items_allocation_len, &dst->keypaths)) != WALLY_OK) {
+                return ret;
+            }
+        }
+
+        for (i = 0; i < src->keypaths->num_items; ++i) {
+            bool found = false;
+            for (j = 0; j < dst->keypaths->num_items; ++j) {
+                if (memcmp((char *)dst->keypaths->items[j].pubkey, (char *)src->keypaths->items[i].pubkey, 65) == 0) {
+                    found = true;
+                    break;
+                }
+            }
+            if (found) {
+                continue;
+            }
+            if ((ret = add_keypath_item(dst->keypaths, &src->keypaths->items[i])) != WALLY_OK) {
+                return ret;
+            }
+        }
+    }
+
+    if (src->unknowns) {
```

Comment:
This block of code copying the unknowns_map seems to be repeated in three places, I wonder if extracting it out into a separate function would be better.

Reply:
Extracted into it's own function. Also did the same to keypaths_map too.